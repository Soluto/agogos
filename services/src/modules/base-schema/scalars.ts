import { gql } from 'apollo-server-core';
import {
  EmailAddressTypeDefinition,
  EmailAddressResolver,
  PhoneNumberTypeDefinition,
  PhoneNumberResolver,
} from 'graphql-scalars';
import GraphQLJSON, { GraphQLJSONObject } from 'graphql-type-json';
import { GraphQLDate, GraphQLDateTime, GraphQLTime } from 'graphql-iso-date';
import { GraphQLResolverMap } from 'apollo-graphql';

export const sdl = gql`
  scalar JSON
  scalar JSONObject

  scalar Date
  scalar Time
  scalar DateTime

  ${EmailAddressTypeDefinition}
  ${PhoneNumberTypeDefinition}
`;

export const resolvers: GraphQLResolverMap = {
  JSON: GraphQLJSON,
  JSONObject: GraphQLJSONObject,
  Date: GraphQLDate,
  Time: GraphQLTime,
  DateTime: GraphQLDateTime,
  EmailAddress: EmailAddressResolver,
  PhoneNumber: PhoneNumberResolver,
};
