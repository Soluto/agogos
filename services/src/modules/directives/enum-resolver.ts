import { gql } from 'apollo-server-core';
import { EnumValueNode, GraphQLEnumType, GraphQLEnumValue, ValueNode } from 'graphql';
import { SchemaDirectiveVisitor } from 'graphql-tools';
import { GraphQLJSONObject } from 'graphql-scalars';

function parseValueNode(valueNode: ValueNode) {
  switch (valueNode.kind) {
    case 'BooleanValue':
    case 'IntValue':
    case 'FloatValue':
    case 'StringValue':
    case 'EnumValue':
      return valueNode.value;

    case 'ObjectValue':
      return GraphQLJSONObject.parseLiteral(valueNode, {});

    default:
      throw new Error(`Unexpected value node type ${valueNode.kind}`);
  }
}

export class EnumValueDirective extends SchemaDirectiveVisitor {
  visitEnumValue(value: GraphQLEnumValue) {
    return value;
  }
}

export class EnumResolverDirective extends SchemaDirectiveVisitor {
  visitEnum(type: GraphQLEnumType): GraphQLEnumType | void | null {
    const values = type.getValues().map(enumVal => {
      const key = enumVal.name;
      const enumValueDirective = enumVal.astNode?.directives?.find(d => d.name.value === 'enumValue');
      if (!enumValueDirective) {
        throw new Error('Each ENUM_VALUE must have @enumValue directive');
      }
      const valNode = enumValueDirective.arguments?.find(a => a.name.value === 'value')?.value;
      if (!valNode) {
        throw new Error('@enumValue directive must have "value" argument');
      }
      const value = parseValueNode(valNode);
      return { key, value };
    });

    type.parseValue = k => values.find(({ key }) => key === k)?.value;
    type.serialize = v => values.find(({ value }) => value === v)?.key;
    type.parseLiteral = keyNode => {
      const k = (keyNode as EnumValueNode).value;
      return values.find(({ key }) => key === k)?.value;
    };
  }
}

export const sdl = gql`
  directive @enumValue(value: JSON!) on ENUM_VALUE
  directive @enumResolver on ENUM
`;
