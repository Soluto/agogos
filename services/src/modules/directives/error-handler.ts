import { gql } from 'apollo-server-core';
import { GraphQLFieldResolverParams } from 'apollo-server-types';
import { defaultFieldResolver, FieldNode, GraphQLField, GraphQLObjectType, GraphQLResolveInfo } from 'graphql';
import { SchemaDirectiveVisitor } from 'graphql-tools';
import { injectArgs } from '../arguments-injection';
import { RequestContext } from '../context';

const defaultObjectResolver = (
  source: unknown,
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  _fields: Record<string, ReadonlyArray<FieldNode>>,
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  _context: RequestContext,
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  _info: GraphQLResolveInfo
) => source;

export class ErrorHandlerDirective extends SchemaDirectiveVisitor {
  visitObject(object: GraphQLObjectType<unknown, RequestContext>) {
    const originalResolve = object.resolveObject || defaultObjectResolver;
    const { catchError, throwError } = this.args as ErrorHandlerDirectiveArgs;

    object.resolveObject = async (source, fields, context, info) => {
      const injectionArgs = { source, args: {}, context, info };
      let result;
      try {
        result = await originalResolve.call(object, source, fields, context, info);
      } catch (error) {
        result = handleCatchErrorArgument(catchError, error, injectionArgs);
      }
      throwError && handleThrowErrorArgument(throwError, result, injectionArgs);
      return result;
    };
  }

  visitFieldDefinition(field: GraphQLField<unknown, RequestContext>) {
    const originalResolve = field.resolve || defaultFieldResolver;
    const { catchError, throwError } = this.args as ErrorHandlerDirectiveArgs;

    field.resolve = async (source, args, context, info) => {
      const injectionArgs = { source, args, context, info };
      let result;
      try {
        result = await originalResolve.call(field, source, args, context, info);
      } catch (error) {
        result = handleCatchErrorArgument(catchError, error, injectionArgs);
      }

      throwError && handleThrowErrorArgument(throwError, result, injectionArgs);
      return result;
    };
  }
}

export const sdl = gql`
  input CatchErrorInput {
    condition: String
    returnValue: JSON
  }

  input ThrowErrorInput {
    condition: String
    errorToThrow: String
  }

  directive @errorHandler(catchError: CatchErrorInput, throwError: ThrowErrorInput) on OBJECT | FIELD_DEFINITION
`;

type ErrorHandlerDirectiveArgs = {
  catchError?: {
    condition?: string;
    returnValue: string | Record<string, unknown>;
  };
  throwError?: {
    condition?: string;
    errorToThrow: string;
  };
};

function handleCatchErrorArgument(
  catchError: { condition?: string | undefined; returnValue?: string | Record<string, unknown> } | undefined,
  error: Error,
  injectionArgs: GraphQLFieldResolverParams<unknown, RequestContext>
) {
  if (!catchError) throw error;

  let shouldCatch = true;
  if (catchError.condition) {
    const catchErrorConditionResult = injectArgs(catchError.condition, { error, ...injectionArgs });
    // If injection throws it returns the template. In this case fail the condition
    shouldCatch = !!catchErrorConditionResult && catchErrorConditionResult != catchError.condition;
  }
  if (!shouldCatch) {
    throw error;
  }

  let result: unknown;
  if (!catchError.returnValue) {
    result = null;
  } else if (typeof catchError.returnValue === 'string') {
    result = injectArgs(catchError.returnValue, { error, ...injectionArgs });
  } else {
    result = catchError.returnValue;
  }
  return result;
}

function handleThrowErrorArgument(
  throwError: { condition?: string; errorToThrow: string },
  result: unknown,
  injectionArgs: GraphQLFieldResolverParams<unknown, RequestContext>
) {
  let shouldThrow = true;
  if (throwError.condition) {
    const throwErrorConditionResult = injectArgs(throwError.condition, { result, ...injectionArgs });
    // If injection throws it returns the template. In this case fail the condition
    shouldThrow = !!throwErrorConditionResult && throwErrorConditionResult != throwError.condition;
  }
  if (!shouldThrow) return;

  let err = throwError.errorToThrow ? injectArgs(throwError.errorToThrow, { result, ...injectionArgs }) : '';
  if (typeof err === 'string') {
    err = new Error(err);
  }
  throw err;
}
