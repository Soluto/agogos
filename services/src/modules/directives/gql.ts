import { SchemaDirectiveVisitor, makeRemoteExecutableSchema, delegateToSchema, introspectSchema } from 'graphql-tools';
import { GraphQLField, GraphQLSchema } from 'graphql';
import { gql } from 'apollo-server-core';
import { ApolloLink } from 'apollo-link';
import { HttpLink } from 'apollo-link-http';
import { RetryLink } from 'apollo-link-retry';
import { setContext } from 'apollo-link-context';
import fetch from 'node-fetch';
import { FastifyRequest } from 'fastify';
import { inject } from '../arguments-injection';
import { ActiveDirectoryAuth, getAuthHeaders } from '../upstreams/authentication';
import logger from '../logger';
import { RequestContext } from '../context';
import { ResourceGroup } from '../resource-repository';

const pendingIntrospectionRetries = new Map<string, NodeJS.Timeout>();

export class GqlDirective extends SchemaDirectiveVisitor {
  createRemoteSchema(
    url: string,
    resourceGroup: ResourceGroup,
    activeDirectoryAuth: ActiveDirectoryAuth,
    timeoutMs?: number
  ) {
    const httpLink: ApolloLink = new HttpLink({
      uri: url,
      fetch: fetch as any,
      fetchOptions: { timeout: timeoutMs ?? 10000 },
    });
    const authLink = setContext(async (_, context) => ({
      headers: await getAuthHeaders(
        resourceGroup,
        activeDirectoryAuth,
        new URL(url).host,
        context?.graphqlContext?.request as FastifyRequest
      ),
    })).concat(httpLink);
    const retryLink = new RetryLink({
      delay: { max: 5000 },
      attempts: {
        max: 5,
        retryIf(error) {
          logger.warn({ error, url }, 'Failed fetching introspection query');
          return !!error;
        },
      },
    }).concat(authLink);

    pendingIntrospectionRetries.delete(url);

    const result: RemoteSchemaWithStatus = { ready: false };
    this.runIntrospection(retryLink, authLink, result, url);
    return result;
  }

  runIntrospection(retryLink: ApolloLink, authLink: ApolloLink, result: RemoteSchemaWithStatus, url: string) {
    // Only introspection should retry, because if we don't have the introspection result this entire resolver will fail.
    // Normal gql requests should not retry, at least for now, because that is more complicated than introspection.
    introspectSchema(retryLink)
      .then(schema => makeRemoteExecutableSchema({ schema, link: authLink }))
      .then(schema => Object.assign(result, { schema, ready: true }))
      .catch(err => {
        logger.error({ err, url }, 'Failed all retries of fetching introspection query, will retry in 10 minutes');
        // https://github.com/Microsoft/TypeScript/issues/30128
        const timeout = global.setTimeout(() => {
          this.runIntrospection(retryLink, authLink, result, url);
        }, 10 * 60 * 1000);
        pendingIntrospectionRetries.set(url, timeout);
      });
  }

  visitFieldDefinition(field: GraphQLField<unknown, RequestContext>) {
    const { url, fieldName, arguments: gqlArgs, operationType: operationTypeParam, timeoutMs } = this.args;
    const operationType = operationTypeParam?.toLowerCase() ?? 'query';
    const remoteSchema = this.createRemoteSchema(
      url,
      this.context.resourceGroup,
      this.context.activeDirectoryAuth,
      timeoutMs
    );

    field.resolve = async (source, args, context, info) => {
      if (!remoteSchema.ready) throw new Error(`Failed reaching remote gql server for introspection (url ${url})`);

      return await delegateToSchema({
        schema: remoteSchema.schema!,
        operation: operationType,
        fieldName,
        args: inject(gqlArgs, { source, args, context, info }) as Record<string, unknown>,
        context,
        info,
      });
    };
  }
}

type RemoteSchemaWithStatus = {
  ready: boolean;
  schema?: GraphQLSchema;
};

export const sdl = gql`
  enum GraphQLOperationType {
    Query
    Mutation
  }

  directive @gql(
    url: String!
    fieldName: String!
    operationType: GraphQLOperationType
    arguments: JSONObject
    timeoutMs: Int
  ) on FIELD_DEFINITION
`;
