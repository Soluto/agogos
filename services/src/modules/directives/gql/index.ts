import { sdl, GqlDirective } from './directive';
import { updateRemoteGqlSchemas, RemoteSchema } from './introspection';

export { sdl, GqlDirective, updateRemoteGqlSchemas, RemoteSchema };
