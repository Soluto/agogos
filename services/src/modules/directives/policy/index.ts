import { PolicyDefinition } from '../../resource-repository';
import { sdl as policySdl, PolicyDirective } from './policy';
import { sdl as policiesSdl, PoliciesDirective, resolvers as policyScalarResolvers } from './policies';
import { sdl as policyQuerySdl, PolicyQueryDirective } from './policy-query';
import PolicyExecutor from './policy-executor';
import { Policy, LoadedPolicy } from './types';

export {
  PolicyExecutor,
  policySdl,
  policiesSdl,
  policyQuerySdl,
  PolicyDirective,
  PoliciesDirective,
  PolicyQueryDirective,
  policyScalarResolvers,
};

export type AuthorizationConfig = {
  policies: PolicyDefinition[];
  policyAttachments?: Record<string, LoadedPolicy>;
  policyExecutor: PolicyExecutor;
  basePolicy?: Policy;
};
