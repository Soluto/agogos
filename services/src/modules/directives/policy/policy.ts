import { GraphQLResolveInfo, GraphQLField, defaultFieldResolver, GraphQLObjectType, FieldNode } from 'graphql';
import { SchemaDirectiveVisitor } from 'graphql-tools';
import { gql } from 'apollo-server-core';
import { RequestContext } from '../../context';
import { GraphQLArguments, Policy } from './types';

export class PolicyDirective extends SchemaDirectiveVisitor {
  visitObject(object: GraphQLObjectType<unknown, RequestContext>) {
    const policy = this.args as Policy;
    const originalResolveObject = object.resolveObject;

    object.resolveObject = async (
      source: unknown,
      fields: Record<string, ReadonlyArray<FieldNode>>,
      context: RequestContext,
      info: GraphQLResolveInfo
    ) => {
      if (!context.ignorePolicies) {
        await context.authorizationConfig.policyExecutor.validatePolicy(policy, source, {}, context, info);
      }

      return originalResolveObject ? await originalResolveObject(source, fields, context, info) : source;
    };
  }

  visitFieldDefinition(field: GraphQLField<unknown, RequestContext>) {
    const policy = this.args as Policy;
    const originalResolve = field.resolve || defaultFieldResolver;

    field.resolve = async (
      source: unknown,
      args: GraphQLArguments,
      context: RequestContext,
      info: GraphQLResolveInfo
    ) => {
      if (!context.ignorePolicies) {
        await context.authorizationConfig.policyExecutor.validatePolicy(policy, source, args, context, info);
      }

      return originalResolve.call(field, source, args, context, info);
    };
  }
}

export const sdl = gql`
  directive @policy(namespace: String!, name: String!, args: JSONObject) on OBJECT | FIELD_DEFINITION
`;
