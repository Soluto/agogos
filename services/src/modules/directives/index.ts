import { concatAST } from 'graphql';
import { SchemaDirectiveVisitor } from 'graphql-tools';
import { sdl as stubSdl, StubDirective } from './stub';
import { sdl as localResolverSdl, LocalResolverDirective } from './local-resolver';
import { sdl as restSdl, RestDirective } from './rest';
import { sdl as gqlSdl, GqlDirective } from './gql';
import { sdl as exportSdl, ExportDirective } from './export';
import { sdl as selectSdl, SelectDirective } from './select';
import {
  policyBaseSdl,
  policySdl,
  policiesSdl,
  policyQuerySdl,
  PolicyDirective,
  PoliciesDirective,
  PolicyQueryDirective,
  policyScalarResolvers,
} from './policy';

export const directiveMap: { [visitorName: string]: typeof SchemaDirectiveVisitor } = {
  stub: StubDirective,
  rest: RestDirective,
  gql: GqlDirective,
  export: ExportDirective,
  select: SelectDirective,
  policy: PolicyDirective,
  policies: PoliciesDirective,
  policyQuery: PolicyQueryDirective,
  localResolver: LocalResolverDirective,
};

export const sdl = concatAST([
  stubSdl,
  restSdl,
  gqlSdl,
  exportSdl,
  selectSdl,
  policySdl,
  policiesSdl,
  policyQuerySdl,
  localResolverSdl,
]);

export { policyScalarResolvers, policyBaseSdl };
