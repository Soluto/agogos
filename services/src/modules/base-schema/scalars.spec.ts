import { concatAST } from 'graphql';
import { gql, ApolloServerBase } from 'apollo-server-core';
import { ApolloServer } from 'apollo-server-fastify';
import { createTestClient, ApolloServerTestClient } from 'apollo-server-testing';
import { policiesSdl } from '../directives/policy';
import { sdl, resolvers } from './scalars';

const testCases: [string, unknown][] = [
  ['JSON', ['bar', 'baz']],
  ['JSONObject', { bar: 1, baz: true }],
  ['Date', new Date(Date.UTC(2020, 11, 24))],
  ['Time', new Date(Date.UTC(2017, 0, 10, 14, 30))],
  ['DateTime', new Date(Date.UTC(2017, 0, 10, 21, 33, 15, 233))],
  ['PhoneNumber', '+12225555'],
  ['EmailAddress', 'stitch@email.com'],
  ['PolicyDetails', { namespace: 'ns', name: 'my-policy', args: { foo: '{source.foo}' } }],
];

describe('Scalars', () => {
  const createTestSchema = (scalar: string) => gql`
    type Query {
      foo: ${scalar}!
    }
  `;
  const createTestResolver = <T>(data: T) => ({
    Query: {
      foo: () => data,
    },
  });

  const query = gql`
    query {
      foo
    }
  `;

  let client: ApolloServerTestClient;
  let server: ApolloServerBase;

  test.each(testCases)('%s is declared', async (scalarName, scalarValue) => {
    server = new ApolloServer({
      typeDefs: concatAST([createTestSchema(scalarName), sdl, policiesSdl]),
      resolvers: [createTestResolver(scalarValue), resolvers],
    });
    client = createTestClient(server);

    const { data, errors } = await client.query({ query });
    expect({ data, errors }).toMatchSnapshot();
  });
});
