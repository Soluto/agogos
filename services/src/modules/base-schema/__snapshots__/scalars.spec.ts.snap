// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Scalars Date is declared 1`] = `
Object {
  "data": Object {
    "foo": "2020-12-23",
  },
  "errors": undefined,
}
`;

exports[`Scalars DateTime is declared 1`] = `
Object {
  "data": Object {
    "foo": "2017-01-10T21:33:15.233Z",
  },
  "errors": undefined,
}
`;

exports[`Scalars EmailAddress is declared 1`] = `
Object {
  "data": Object {
    "foo": "stitch@email.com",
  },
  "errors": undefined,
}
`;

exports[`Scalars JSON is declared 1`] = `
Object {
  "data": Object {
    "foo": Array [
      "bar",
      "baz",
    ],
  },
  "errors": undefined,
}
`;

exports[`Scalars JSONObject is declared 1`] = `
Object {
  "data": Object {
    "foo": Object {
      "bar": 1,
      "baz": true,
    },
  },
  "errors": undefined,
}
`;

exports[`Scalars PhoneNumber is declared 1`] = `
Object {
  "data": Object {
    "foo": "+12225555",
  },
  "errors": undefined,
}
`;

exports[`Scalars Time is declared 1`] = `
Object {
  "data": Object {
    "foo": "14:30:00.000Z",
  },
  "errors": undefined,
}
`;
